File: 백준/Gold/7569. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    static class Node {
        int h, x, y, cnt;
        public Node(int h, int x, int y, int cnt) {
            this.h = h;
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int M = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        int H = Integer.parseInt(st.nextToken());
        int[][][] map = new int[H][N][M];
        Queue<Node> que = new LinkedList<>();
        int[] dx = {-1, 1, 0, 0, 0, 0};
        int[] dy = {0, 0, -1, 1, 0, 0};
        int[] dh = {0, 0, 0, 0, -1, 1};
        int need = 0;
        for(int h = 0; h < H; h++) {
            for(int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                for(int j = 0; j < M; j++) {
                    map[h][i][j] = Integer.parseInt(st.nextToken());
                    if(map[h][i][j] == 0) need++;
                    else if(map[h][i][j] == 1) que.add(new Node(h, i, j, 0));
                }
            }
        }
        if(need == 0) {
            System.out.println(0);
            return;
        }
        while(!que.isEmpty()) {
            Node node = que.poll();
            for(int i = 0; i < 6; i++) {
                int nh = node.h + dh[i];
                int nx = node.x + dx[i];
                int ny = node.y + dy[i];
                if(nh < 0 || nx < 0 || ny < 0 || nh >= H || nx >= N || ny >= M || map[nh][nx][ny] != 0) continue;
                que.add(new Node(nh, nx, ny, node.cnt + 1));
                map[nh][nx][ny] = 1;
                need--;
                if(need == 0) {
                    System.out.println(node.cnt + 1);
                    return;
                }
            }
        }
        System.out.println(-1);
    }
}

File: 백준/Gold/7569. 토마토/토마토.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.arr.length === this.head;
  }
}
const [[M, N, H], ...arr] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
let need = 0;
const map = [];
for(let h = 0; h < H; h++) {
  map.push(arr.slice(h * N, (h + 1) * N));
}
const que = new Queue();
const dh = [-1, 1, 0, 0, 0, 0];
const dx = [0, 0, -1, 1, 0, 0];
const dy = [0, 0, 0, 0, -1, 1];
for(let h = 0; h < H; h++) {
  for(let i = 0; i < N; i++) {
    for(let j = 0; j < M; j++) {
      if(map[h][i][j] === 0) need++;
      else if(map[h][i][j] === 1) que.add({h : h, x : i, y : j, cnt : 0});
    }
  }
}
if(need === 0) {
  console.log(0);
  return;
}
while(!que.isEmpty()) {
  const node = que.poll();
  for(let i = 0; i < 6; i++) {
    const nh = node.h + dh[i];
    const nx = node.x + dx[i];
    const ny = node.y + dy[i];
    if(nh < 0 || nx < 0 || ny < 0 || nh >= H || nx >= N || ny >= M || map[nh][nx][ny] != 0) continue;
    que.add({h : nh, x : nx, y : ny, cnt : node.cnt + 1});
    map[nh][nx][ny] = 1;
    need--;
    if(need === 0) {
      console.log(node.cnt + 1);
      return;
    }
  }
}
console.log(-1);